### ISSUE TITLE
FIRM payload

### ISSUE THREAD

**Issue by chaoskagami** [chaoskagami/corbenik#72]
Corbenik should be buildable as a firm payload. This entails a bit more work than I'd like, but time to strap down and get to work I guess.

 - [x] Build firm payload
 - [x] FIX CRASH (fml) <- not really fixed, worked around for now
 - [ ] Figure out why TWL/AGB won't properly decrypt now (keys?)
 - [ ] Add support for chainloading firm payloads
 - [ ] Add a fallback mode for when no SD is inserted

**Comment by gnmmarechal** [comment]
Would it be remotely feasible to make Corbenik run older, non FIRM payloads? I cry when I think about backwards compatibility being gone

**Comment by chaoskagami** [comment]
I'm keeping support for raw arm9 binaries.

**Comment by gnmmarechal** [comment]
yay

**Comment by chaoskagami** [comment]
For now, I just want to get task #1 done so it can actually be booted via boot9strap. Everything else can come later.

**Comment by gnmmarechal** [comment]
So, launching Corbenik from B9S will allow me to run an older ARM9 payload (I mean, when you're done)?

**Comment by chaoskagami** [comment]
.bin = raw arm9 binary

**Comment by chaoskagami** [comment]
Wrong button.

**Comment by gnmmarechal** [comment]
Just making sure I wasn't misunderstanding. And, will Corbenik be FIRM-only, or will builds build both? Or will it be FIRM by default, and ARM9 version will be built only if needed?

**Comment by chaoskagami** [comment]
Unless I start to add boot9-specific functionality, both will be supported. For now.

This isn't an excuse to not switch from arm9loaderhax, however. boot9 is a much cleaner exploit. I also prefer to keep #ifdef'ing to a minimum.

**Comment by chaoskagami** [comment]
And to be clear; you do NOT want to flash corbenik to FIRM0.

**Comment by gnmmarechal** [comment]
yep, I already upgraded my 2DS, I'll upgrade my 3DS once Corbenik is functional there.

**Comment by Derek1199** [comment]
Will Corbenik support b9strap soon?

**Comment by chaoskagami** [comment]
Within a week or two? Sure. If you mean "FIX NAOW", then no. Definitely not. I have a real life and the obligations that come with it.

**Comment by Derek1199** [comment]
Oh, OK. Sounds good. And yeah a week or two isn't too bad. Lol. I can wait
that long. Anyway, I wish you good luck. Lol. Your a great dev :v

**Comment by NinjaBoyLao** [comment]
i'll gladly help you test stuff out for breakage

**Comment by Lordkurzschluss** [comment]
@chaoskagami One Question: is Corbenik dead? Many people miss it.
Do you have any plans for finishing it on B9s?

**Comment by gnmmarechal** [comment]
Corbenik's development has been halted.

Refer to http://gbatemp.net/threads/corbenik-another-cfw-for-advanced-users-with-bytecode-patches.429612/page-105#post-7371367

**Comment by d3m3vilurr** [comment]
@Arcterus forked a project. https://github.com/Arcterus/corbenik

**Comment by chaoskagami** [comment]
No commits since June 10 (about a month ago), sadly.

That said, I approve of what he's done so far. Quick list for those who don't feel like reading the code he changed:
 * FIRM chainloading support
 * Interpreter rewritten to use computed goto (which is theoretically a major performance increase over switch)
 * Screen buffering improvements

**Comment by ghost** [comment]
Hi @chaoskagami long time no see. :)

I wanted to ask you about 3DS CFW Patches. How exactly do I put them in a CFW? What stuff's needed?

I'm currently making a CFW and was wondering if maybe you could tell me what to use?

**Comment by ghost** [comment]
@chaoskagami mainly the pco patches.
